### DISCUSSION TITLE
Set-OutlookSignatures.ps1 unable to complete successfully for standard users (works perfectly for those with admin rights)

### DISCUSSION THREAD

**Discussion by CarlInLV** [Set-OutlookSignatures/Set-OutlookSignatures#114] [Q&A] [answered]
Hello,
Thank you for creating and sharing this very helpful utility!
I ran the Set-OutlookSignatures.ps1 while logged in with an account that has Domain Administrator permissions and encountered no issues.
I created custom _Signatures.ini, replacement variable config file, and graph config file.
When I launch this from my own PC, it works great and I can immediately see and use the signature based on the template (we're only standardizing one signature for now).
powershell.exe -noexit -file "X:\Redacted\Set-OutlookSignatures\Set-OutlookSignatures.ps1" -GraphOnly true -SignatureTemplatePath 'X:\Redacted\Set-OutlookSignatures\Redacted Templates\Signatures DOCX' -SignatureIniPath 'X:\Redacted\Set-OutlookSignatures\Redacted Templates\Signatures DOCX\_RedactedSignatures.ini' -ReplacementVariableConfigFile "X:\Redacted\Set-OutlookSignatures\Redactedconfig\Redacted replacement variables.ps1" -GraphConfigFile "X:\Redacted\Set-OutlookSignatures\Redactedconfig\Redacted graph config.ps1"

Now when I launch that same script while logged in as a help desk user with slightly elevated permissions (local admin on the local workstation) or if the script launches as per the scheduled task (which runs the script at any user logon), the end result is a window pops up that says "You started Set-OutlookSignatures, or an Administrator configured it to run for you to update your Outlook signatures and out-of-office replies. A required security token for access to Microsoft 365 is not yet available..."

The last paragraph says to login to Micrososft 365 with my account in the new browser tab that will open after I close this message, but the browser never opens after I close that message so I can't figure out how to work through this.

This is a Windows environment (clients are on Windows 10 Enterprise and servers are 2016/2019/2022 Standard/Enterprise) with on-prem A.D. which syncs to M365/Azure (we're hybrid configured). We do not have a local MS Exchange server. All of our mailboxes are in Microsoft's Cloud. *edit 3* - All of our workstations are equipped with Microsoft Office 2019 Professional Plus.

My intent is to use Set-OutlookSignatures to apply one standardized email signature to all users (all of whom have a mailbox with Microsoft 365).

Please let me know what else to try, or what other pieces of information I may provide to assist. Thanks again!

*edit* - I have manually set the execution policy within PowerShell to bypass on the PC I am testing with. I do plan on setting this back to Restricted once successful testing occurs. I have also used Group Policy to allow Google Chrome (it's our default browser) to "allow insecure content on these sites" = localhost

*edit 2* - the only text that appears in red when I launch the script as a standard user (or as the help desk user) is "Problem connecting to Microsoft Graph. Exit." - This didn't happen to me, I assume because I have Global Administrator rights within our M365 tenant.

**Comment by GruberMarkus** [comment]
Dear @CarlInLV ,

here are may thoughts on your issue:

**Popup informing the user about authentication**
The popup appears because a prior cached authentication to Microsoft Graph is no longer valid, or because silent authentication has failed (including Windows Integrated Authentication).

The text of the popup can be modified via a custom graph config file. I would not disable it, as users else wonder why an unknown application asks for authentication (due to a Microsoft bug, the name of the application is not shown in Graph authentication, this is unfixed for more than a year).

If you want to know why Windows Integrated Authentication and silent authentication fails, you need to run Set-OutlookSignatures with the "-verbose" parameter.

**Error message from Edit 2**
You should not only see "Problem connecting to Microsoft Graph. Exit.", but also the following:
```
No Graph authentication possible.
1. Did you follow the Quick Start Guide in '.\docs\README' and configure the Entra ID/Azure AD app correctly?
2. If the "Via Prompt with LoginHint and Timeout" authentication message is diplayed:
   - Does a browser (the system default browser, if configured) open and ask for authentication?
     - Yes:
       - Check if the correct user account is selected/entered and if the authentication is successful
       - Check if authentication happens within two minutes
       - Ensure that access to 'http://localhost' is allowed ('https://localhost' is currently not technically feasible, see 'https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/System-Browser-on-.Net-Core' and 'https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/MSAL.NET-uses-web-browser' for details)
     - No:
       - Run Set-OutlookSignatures in a new PowerShell session
       - Check if a default browser is set and if "start https://github.com/Set-OutlookSignatures/Set-OutlookSignatures" opens it
       - Make sure that Set-OutlookSignatures is executed in the security context of the currently logged-in user
       - Make sure that the current PowerShell session allows TLS 1.2+ (see https://github.com/Set-OutlookSignatures/Set-OutlookSignatures/issues/85 for details)
3. Run Set-OutlookSignatures with the "-Verbose" parameter and check for authentication messages
4. Delete the MSAL.PS Graph token cache: $($script:msalClientApp.cacheInfo).
```

You write that no browser window opens. This is a very strong hint to the likely root cause of the problem. The test you should perform is mentioned in the error message: `Check if a default browser is set and if "start https://github.com/Set-OutlookSignatures/Set-OutlookSignatures" opens it`

What does this command do?

**Reply by CarlInLV** [reply]
Hello @GruberMarkus!
I very much appreciate how quickly you replied! Thank you!
I do not intend on disabling the popup, my intent is to successfully test the utility before rolling it out to all users (will roll out to IT first).
I ran Set-OutlookSignatures with the -verbose parameter.
I do see the text after the "No Graph authentication possible" message and am working through the "if no" portion.

I did follow the quick start guide, and I believe I configured the Entra ID/Azure AD app correctly because Set-OutlookSignatures works for myself. Is it possible to have the utility work correctly only for myself if the Entra ID/Azure AD app was not configured correctly?

In a separate PowerShell window, I ran the "start https://github.com/Set-OutlookSignatures/Set-OutlookSignatures" command and confirmed this did not launch the browser.

Thank you for pointing me in the right direction, I will continue troubleshooting intent on getting that command to open the correct browser and will go from there.

**Reply by CarlInLV** [reply]
The command not opening the browser might have been due to how I input it. When I run the command "start https://github.com/Set-OutlookSignatures/Set-OutlookSignatures" (without the quotes), Chrome does open to the Set-OutlookSignatures Github page.

**Comment by GruberMarkus** [comment]
Yes, it is possible to configure an Entra ID app so that it is only working for some specific users. For example, when you grant consent only for your own admin user, but not for the whole tenant. But I don't think that this is your problem.

Your message `In a separate PowerShell window, I ran the "start https://github.com/Set-OutlookSignatures/Set-OutlookSignatures" command and confirmed this did not launch the browser.` is the hint we needed: Your test user does not have configured a default browser for the https protocol.

Your test confirms that the root cause is not in Set-OutlookSignatures, but that it is a general problem for the logged-on user.

**Comment by GruberMarkus** [comment]
In an edit, you write that Chrome does open. This would mean that a default browser listenting to the https protocol is configured.

Please run Set-OutlookSignatures with the "-verbose" parameter and provide the full log.

Please do not use this discussion, but create an issue at https://github.com/Set-OutlookSignatures/Set-OutlookSignatures/issues.

I will close this discussion after the issue has been created.

**Comment by GruberMarkus** [comment, marked as answer]
Converted to issue #116.
